src/main/java/gregtech/common/items/behaviors/MultiAssemblerBehaviour.java

```diff
@@ -0,0 +1,128 @@
+package gregtech.common.items.behaviors;
+
+import gregtech.api.block.machines.BlockMachine;
+import gregtech.api.items.metaitem.stats.IItemBehaviour;
+import gregtech.api.metatileentity.MetaTileEntity;
+import gregtech.api.metatileentity.MetaTileEntityHolder;
+import gregtech.api.metatileentity.multiblock.MultiblockControllerBase;
+import gregtech.api.multiblock.BlockPatternChecker;
+import gregtech.api.net.CPacketMultiBlockStructure;
+import gregtech.api.net.NetworkHandler;
+import gregtech.api.render.scene.WorldSceneRenderer;
+import gregtech.common.render.MultiHelperRenderer;
+import gregtech.integration.jei.GTJeiOptional;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.resources.I18n;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.util.ActionResult;
+import net.minecraft.util.EnumActionResult;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.text.TextComponentTranslation;
+import net.minecraft.world.World;
+import net.minecraftforge.fml.common.Loader;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class MultiAssemblerBehaviour implements IItemBehaviour {
+
+    @Override
+    public EnumActionResult onItemUseFirst(EntityPlayer player, World world, BlockPos pos, EnumFacing side, float hitX, float hitY, float hitZ, EnumHand hand) {
+        NBTTagCompound nbt = player.getHeldItem(hand).getOrCreateSubCompound("GT.Detrav");
+        byte mode = nbt.hasKey("mode") ? nbt.getByte("mode") : 0;
+        if (pos != null && world.getTileEntity(pos) instanceof MetaTileEntityHolder) {
+            MetaTileEntity mte = ((MetaTileEntityHolder) world.getTileEntity(pos)).getMetaTileEntity();
+            if (mte instanceof MultiblockControllerBase && !player.isSneaking()) { // preview
+                if (world.isRemote) { // client side only
+                    if(!Loader.isModLoaded("jei")) { // check jei installed
+                        player.sendMessage(new TextComponentTranslation("metaitem.multi_assembler.jei_missing"));
+                    } else if (mode == 3) { // building mode
+                        WorldSceneRenderer worldSceneRenderer = GTJeiOptional.getWorldSceneRenderer((MultiblockControllerBase) mte);
```

Comment:
Won't this crash when JEI is not installed? The annotation you used removes the method from the class when JEI is not installed.

It would be better to have an interface (that doesn't depend on JEI) that the JEI integration can implement and initialise a static instance in ClientProxy.

Reply:
I thought about the second point, but I tried to avoid making too many changes to the original code.

Reply:
I would recommed creating an interface called something like MultiblockRenderSource 
that the JEI integration can implement when it is present.

Reply:
good idea